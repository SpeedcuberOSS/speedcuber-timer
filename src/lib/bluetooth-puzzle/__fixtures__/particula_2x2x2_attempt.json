{
  "id": "4aae23b1-4b0a-496f-8d36-27f6cb85defb",
  "extensions": [
    {
      "id": "org.speedcuber.stif.extensions.message_stream",
      "specUrl": "https://speedcuber.org/stif/extensions/message_stream.schema.json",
      "data": {
        "smartPuzzle": {
          "prefix": "GoCube2x2",
          "brand": "Particula GoCube",
          "puzzle": {
            "id": "222"
          },
          "uuids": {
            "trackingService": "6e400001-b5a3-f393-e0a9-e50e24dcca9e",
            "trackingCharacteristic": "6e400003-b5a3-f393-e0a9-e50e24dcca9e"
          }
        },
        "stream": [
          {
            "t": 1675144649193,
            "m": "KgYBCAlCDQo="
          },
          {
            "t": 1675144649703,
            "m": "KgYBAgY5DQo="
          },
          {
            "t": 1675144650244,
            "m": "KgYBBAM4DQo="
          },
          {
            "t": 1675144650454,
            "m": "KgYBAgk8DQo="
          },
          {
            "t": 1675144650754,
            "m": "KgYBBQA2DQo="
          },
          {
            "t": 1675144650853,
            "m": "KgYBCAA5DQo="
          },
          {
            "t": 1675144651023,
            "m": "KgYBBAM4DQo="
          },
          {
            "t": 1675144651053,
            "m": "KgYBAwY6DQo="
          },
          {
            "t": 1675144651234,
            "m": "KgYBBQA2DQo="
          },
          {
            "t": 1675144651564,
            "m": "KgYBCAM8DQo="
          },
          {
            "t": 1675144651655,
            "m": "KgYBBAM4DQo="
          },
          {
            "t": 1675144651748,
            "m": "KgYBAwM3DQo="
          },
          {
            "t": 1675144651850,
            "m": "KgYBAwA0DQo="
          },
          {
            "t": 1675144651954,
            "m": "KgYBBQA2DQo="
          },
          {
            "t": 1675144653394,
            "m": "KgYBCQA6DQo="
          },
          {
            "t": 1675144653544,
            "m": "KgYBCQlDDQo="
          },
          {
            "t": 1675144654266,
            "m": "KgYBBQk/DQo="
          },
          {
            "t": 1675144654325,
            "m": "KgYBBQY8DQo="
          },
          {
            "t": 1675144654415,
            "m": "KgYBCQZADQo="
          },
          {
            "t": 1675144654594,
            "m": "KgYBBAk+DQo="
          },
          {
            "t": 1675144654625,
            "m": "KgYBBAA1DQo="
          },
          {
            "t": 1675144654846,
            "m": "KgYBCQM9DQo="
          },
          {
            "t": 1675144654983,
            "m": "KgYBCQA6DQo="
          },
          {
            "t": 1675144655644,
            "m": "KgYBAwk9DQo="
          },
          {
            "t": 1675144655735,
            "m": "KgYBAwY6DQo="
          },
          {
            "t": 1675144655854,
            "m": "KgYBCQlDDQo="
          },
          {
            "t": 1675144655982,
            "m": "KgYBAgk8DQo="
          },
          {
            "t": 1675144656037,
            "m": "KgYBAgAzDQo="
          },
          {
            "t": 1675144656214,
            "m": "KgYBCAA5DQo="
          }
        ]
      }
    }
  ],
  "unixTimestamp": 1675144636354,
  "inspectionCompleteTimestamp": 1675144648997,
  "event": {
    "id": "222",
    "puzzle": {
      "id": "222"
    },
    "type": "WCA"
  },
  "duration": 7561,
  "solutions": [
    {
      "id": "986b068f-38e4-4e8b-8bee-adfeb1368596",
      "scramble": {
        "id": "3cc13ce1-1a99-4480-9895-657edfd47248",
        "puzzle": {
          "id": "222"
        },
        "algorithm": {
          "id": "b43a935b-ea64-49b6-ae1b-083e57d0044a",
          "moves": [
            "R2",
            "F2",
            "R2",
            "F",
            "U2",
            "F",
            "R2",
            "U",
            "F2",
            "R2",
            "U'",
            "R"
          ]
        },
        "provider": {
          "id": "org.speedcuber.scramblerproviders.mandy",
          "url": "https://scrambleproviders.speedcuber.org/mandy"
        }
      }
    }
  ],
  "infractions": [],
  "comment": ""
}