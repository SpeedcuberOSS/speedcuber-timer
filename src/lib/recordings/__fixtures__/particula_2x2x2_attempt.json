{
  "id": "4aae23b1-4b0a-496f-8d36-27f6cb85defb",
  "inspectionStart": 1675144636354,
  "timerStart": 1675144648997,
  "timerStop": 1675144656558,
  "event": { "id": "222", "puzzles": ["222"], "type": "official" },
  "solutions": [
    {
      "puzzle": "222",
      "scramble": [
        "R2",
        "F2",
        "R2",
        "F",
        "U2",
        "F",
        "R2",
        "U",
        "F2",
        "R2",
        "U'",
        "R"
      ],
      "reconstruction": [
        {
          "label": "unknown",
          "moves": [
            { "m": "R", "t": 196 },
            { "m": "F", "t": 706 },
            { "m": "U", "t": 1247 },
            { "m": "F", "t": 1457 },
            { "m": "U'", "t": 1757 },
            { "m": "R", "t": 1856 },
            { "m": "U", "t": 2026 },
            { "m": "F'", "t": 2056 },
            { "m": "U'", "t": 2237 },
            { "m": "R", "t": 2567 },
            { "m": "U", "t": 2658 },
            { "m": "F'", "t": 2751 },
            { "m": "F'", "t": 2853 },
            { "m": "U'", "t": 2957 },
            { "m": "R'", "t": 4397 },
            { "m": "R'", "t": 4547 },
            { "m": "U'", "t": 5269 },
            { "m": "U'", "t": 5328 },
            { "m": "R'", "t": 5418 },
            { "m": "U", "t": 5597 },
            { "m": "U", "t": 5628 },
            { "m": "R'", "t": 5849 },
            { "m": "R'", "t": 5986 },
            { "m": "F'", "t": 6647 },
            { "m": "F'", "t": 6738 },
            { "m": "R'", "t": 6857 },
            { "m": "F", "t": 6985 },
            { "m": "F", "t": 7040 },
            { "m": "R", "t": 7217 }
          ]
        }
      ]
    }
  ],
  "infractions": [],
  "comment": ""
}
