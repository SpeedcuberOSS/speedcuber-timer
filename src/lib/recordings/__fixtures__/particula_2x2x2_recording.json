{
  "smartPuzzle": {
    "prefix": "GoCube2x2",
    "brand": "Particula GoCube",
    "puzzle": "222",
    "uuids": {
      "trackingService": "6e400001-b5a3-f393-e0a9-e50e24dcca9e",
      "trackingCharacteristic": "6e400003-b5a3-f393-e0a9-e50e24dcca9e"
    }
  },
  "stream": [
    {
      "t": 1675144649193,
      "m": "KgYBCAlCDQo="
    },
    {
      "t": 1675144649703,
      "m": "KgYBAgY5DQo="
    },
    {
      "t": 1675144650244,
      "m": "KgYBBAM4DQo="
    },
    {
      "t": 1675144650454,
      "m": "KgYBAgk8DQo="
    },
    {
      "t": 1675144650754,
      "m": "KgYBBQA2DQo="
    },
    {
      "t": 1675144650853,
      "m": "KgYBCAA5DQo="
    },
    {
      "t": 1675144651023,
      "m": "KgYBBAM4DQo="
    },
    {
      "t": 1675144651053,
      "m": "KgYBAwY6DQo="
    },
    {
      "t": 1675144651234,
      "m": "KgYBBQA2DQo="
    },
    {
      "t": 1675144651564,
      "m": "KgYBCAM8DQo="
    },
    {
      "t": 1675144651655,
      "m": "KgYBBAM4DQo="
    },
    {
      "t": 1675144651748,
      "m": "KgYBAwM3DQo="
    },
    {
      "t": 1675144651850,
      "m": "KgYBAwA0DQo="
    },
    {
      "t": 1675144651954,
      "m": "KgYBBQA2DQo="
    },
    {
      "t": 1675144653394,
      "m": "KgYBCQA6DQo="
    },
    {
      "t": 1675144653544,
      "m": "KgYBCQlDDQo="
    },
    {
      "t": 1675144654266,
      "m": "KgYBBQk/DQo="
    },
    {
      "t": 1675144654325,
      "m": "KgYBBQY8DQo="
    },
    {
      "t": 1675144654415,
      "m": "KgYBCQZADQo="
    },
    {
      "t": 1675144654594,
      "m": "KgYBBAk+DQo="
    },
    {
      "t": 1675144654625,
      "m": "KgYBBAA1DQo="
    },
    {
      "t": 1675144654846,
      "m": "KgYBCQM9DQo="
    },
    {
      "t": 1675144654983,
      "m": "KgYBCQA6DQo="
    },
    {
      "t": 1675144655644,
      "m": "KgYBAwk9DQo="
    },
    {
      "t": 1675144655735,
      "m": "KgYBAwY6DQo="
    },
    {
      "t": 1675144655854,
      "m": "KgYBCQlDDQo="
    },
    {
      "t": 1675144655982,
      "m": "KgYBAgk8DQo="
    },
    {
      "t": 1675144656037,
      "m": "KgYBAgAzDQo="
    },
    {
      "t": 1675144656214,
      "m": "KgYBCAA5DQo="
    }
  ]
}