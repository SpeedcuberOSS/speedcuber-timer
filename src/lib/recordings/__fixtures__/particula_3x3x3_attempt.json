{
  "id": "6d2a9eaf-2f1e-4797-879b-a525142cf8fc",
  "inspectionStart": 1675142895723,
  "timerStart": 1675142907090,
  "timerStop": 1675142918923,
  "event": { "id": "333", "puzzles": ["333"], "type": "official" },
  "solutions": [
    {
      "puzzle": "333",
      "scramble": [
        "L2",
        "B",
        "L2",
        "D'",
        "F2",
        "D",
        "B'",
        "R",
        "D'",
        "F2",
        "R",
        "B2",
        "U",
        "R2",
        "D'",
        "B'",
        "L2",
        "F",
        "U",
        "B"
      ]
    }
  ],
  "infractions": [],
  "comment": ""
}
