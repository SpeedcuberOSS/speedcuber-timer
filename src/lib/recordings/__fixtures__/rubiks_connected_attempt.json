{
  "id": "bd19f5f6-ea3c-4f9d-beb0-2aa9c41e661f",
  "event": { "id": "333", "puzzles": ["333"], "type": "official" },
  "inspectionStart": 1674667995013,
  "timerStart": 1674668004869,
  "timerStop": 1674668019734,
  "solutions": [
    {
      "puzzle": "333",
      "scramble": [
        "L2",
        "U",
        "F2",
        "L2",
        "B2",
        "L",
        "U'",
        "R2",
        "F",
        "R",
        "B",
        "R",
        "U",
        "L'",
        "U",
        "B",
        "R2",
        "B'",
        "D",
        "L"
      ],
      "reconstruction": [
        {
          "label": "cross",
          "moves": [
            { "m": "L'", "t": 531 },
            { "m": "R'", "t": 616 },
            { "m": "B", "t": 1187 },
            { "m": "D'", "t": 1457 },
            { "m": "L", "t": 1684 },
            { "m": "U'", "t": 1947 },
            { "m": "B'", "t": 2626 }
          ]
        },
        {
          "label": "1st pair",
          "moves": [
            { "m": "F", "t": 3288 },
            { "m": "F", "t": 3378 },
            { "m": "D", "t": 3860 },
            { "m": "F'", "t": 4065 },
            { "m": "D'", "t": 4366 }
          ]
        },
        {
          "label": "2nd pair",
          "moves": [
            { "m": "D'", "t": 5176 },
            { "m": "F'", "t": 5298 },
            { "m": "F'", "t": 5387 },
            { "m": "D", "t": 5506 },
            { "m": "F", "t": 5598 },
            { "m": "U'", "t": 5817 },
            { "m": "F'", "t": 5986 },
            { "m": "U", "t": 6137 }
          ]
        },
        {
          "label": "3rd pair",
          "moves": [
            { "m": "D'", "t": 6412 },
            { "m": "F", "t": 6467 },
            { "m": "D", "t": 6676 },
            { "m": "R'", "t": 7187 },
            { "m": "F'", "t": 7312 },
            { "m": "F'", "t": 7426 },
            { "m": "R", "t": 7547 },
            { "m": "F'", "t": 7726 },
            { "m": "U", "t": 7880 },
            { "m": "F", "t": 8027 },
            { "m": "U'", "t": 8147 }
          ]
        },
        {
          "label": "4th pair",
          "moves": [
            { "m": "R", "t": 8536 },
            { "m": "F'", "t": 8631 },
            { "m": "R'", "t": 8716 },
            { "m": "D", "t": 8986 },
            { "m": "R'", "t": 9136 },
            { "m": "D'", "t": 9269 },
            { "m": "R", "t": 9316 }
          ]
        },
        {
          "label": "OLLCP",
          "moves": [
            { "m": "F'", "t": 9856 },
            { "m": "F'", "t": 10014 },
            { "m": "F", "t": 10465 },
            { "m": "U", "t": 10736 },
            { "m": "F", "t": 11031 },
            { "m": "L", "t": 11116 },
            { "m": "F'", "t": 11207 },
            { "m": "L'", "t": 11296 },
            { "m": "U'", "t": 11514 },
            { "m": "F'", "t": 11687 },
            { "m": "D", "t": 11972 },
            { "m": "R", "t": 12048 },
            { "m": "F", "t": 12196 },
            { "m": "R'", "t": 12296 },
            { "m": "F'", "t": 12346 },
            { "m": "D'", "t": 12498 }
          ]
        },
        {
          "label": "EPLL",
          "moves": [
            { "m": "R'", "t": 13126 },
            { "m": "F", "t": 13186 },
            { "m": "R'", "t": 13247 },
            { "m": "F'", "t": 13401 },
            { "m": "R'", "t": 13608 },
            { "m": "F'", "t": 13669 },
            { "m": "R'", "t": 13890 },
            { "m": "F", "t": 13969 },
            { "m": "R", "t": 14104 },
            { "m": "F", "t": 14146 },
            { "m": "R", "t": 14266 },
            { "m": "R", "t": 14362 }
          ]
        }
      ]
    }
  ],
  "infractions": [],
  "comment": ""
}
