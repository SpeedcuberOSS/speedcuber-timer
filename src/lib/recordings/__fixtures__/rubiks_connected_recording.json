{
  "smartPuzzle": {
    "prefix": "Rubiks",
    "brand": "Rubik's Connected",
    "puzzle": "333",
    "uuids": {
      "trackingService": "6e400001-b5a3-f393-e0a9-e50e24dcca9e",
      "trackingCharacteristic": "6e400003-b5a3-f393-e0a9-e50e24dcca9e"
    }
  },
  "stream": [
    {
      "t": 1674668005400,
      "m": "KgYBCwlFDQo="
    },
    {
      "t": 1674668005485,
      "m": "KgYBCQM9DQo="
    },
    {
      "t": 1674668006056,
      "m": "KgYBAAM0DQo="
    },
    {
      "t": 1674668006326,
      "m": "KgYBBwA4DQo="
    },
    {
      "t": 1674668006553,
      "m": "KgYBCgA7DQo="
    },
    {
      "t": 1674668006816,
      "m": "KgYBBQM5DQo="
    },
    {
      "t": 1674668007495,
      "m": "KgYBAQAyDQo="
    },
    {
      "t": 1674668008157,
      "m": "KgYBAgY5DQo="
    },
    {
      "t": 1674668008247,
      "m": "KgYBAgk8DQo="
    },
    {
      "t": 1674668008729,
      "m": "KgYBBgM6DQo="
    },
    {
      "t": 1674668008934,
      "m": "KgYBAwY6DQo="
    },
    {
      "t": 1674668009235,
      "m": "KgYBBwA4DQo="
    },
    {
      "t": 1674668010045,
      "m": "KgYBBwlBDQo="
    },
    {
      "t": 1674668010167,
      "m": "KgYBAwM3DQo="
    },
    {
      "t": 1674668010256,
      "m": "KgYBAwA0DQo="
    },
    {
      "t": 1674668010375,
      "m": "KgYBBgA3DQo="
    },
    {
      "t": 1674668010467,
      "m": "KgYBAgM2DQo="
    },
    {
      "t": 1674668010686,
      "m": "KgYBBQA2DQo="
    },
    {
      "t": 1674668010855,
      "m": "KgYBAwA0DQo="
    },
    {
      "t": 1674668011006,
      "m": "KgYBBAM4DQo="
    },
    {
      "t": 1674668011281,
      "m": "KgYBBwlBDQo="
    },
    {
      "t": 1674668011336,
      "m": "KgYBAgM2DQo="
    },
    {
      "t": 1674668011545,
      "m": "KgYBBgA3DQo="
    },
    {
      "t": 1674668012056,
      "m": "KgYBCQA6DQo="
    },
    {
      "t": 1674668012181,
      "m": "KgYBAwA0DQo="
    },
    {
      "t": 1674668012295,
      "m": "KgYBAwk9DQo="
    },
    {
      "t": 1674668012416,
      "m": "KgYBCAM8DQo="
    },
    {
      "t": 1674668012595,
      "m": "KgYBAwY6DQo="
    },
    {
      "t": 1674668012749,
      "m": "KgYBBAY7DQo="
    },
    {
      "t": 1674668012896,
      "m": "KgYBAgk8DQo="
    },
    {
      "t": 1674668013016,
      "m": "KgYBBQM5DQo="
    },
    {
      "t": 1674668013405,
      "m": "KgYBCAY/DQo="
    },
    {
      "t": 1674668013500,
      "m": "KgYBAwY6DQo="
    },
    {
      "t": 1674668013585,
      "m": "KgYBCQM9DQo="
    },
    {
      "t": 1674668013855,
      "m": "KgYBBgM6DQo="
    },
    {
      "t": 1674668014005,
      "m": "KgYBCQA6DQo="
    },
    {
      "t": 1674668014138,
      "m": "KgYBBwA4DQo="
    },
    {
      "t": 1674668014185,
      "m": "KgYBCAM8DQo="
    },
    {
      "t": 1674668014725,
      "m": "KgYBAwM3DQo="
    },
    {
      "t": 1674668014883,
      "m": "KgYBAwA0DQo="
    },
    {
      "t": 1674668015334,
      "m": "KgYBAgM2DQo="
    },
    {
      "t": 1674668015605,
      "m": "KgYBBAY7DQo="
    },
    {
      "t": 1674668015900,
      "m": "KgYBAgY5DQo="
    },
    {
      "t": 1674668015985,
      "m": "KgYBCgM+DQo="
    },
    {
      "t": 1674668016076,
      "m": "KgYBAwM3DQo="
    },
    {
      "t": 1674668016165,
      "m": "KgYBCwA8DQo="
    },
    {
      "t": 1674668016383,
      "m": "KgYBBQM5DQo="
    },
    {
      "t": 1674668016556,
      "m": "KgYBAwA0DQo="
    },
    {
      "t": 1674668016841,
      "m": "KgYBBgM6DQo="
    },
    {
      "t": 1674668016917,
      "m": "KgYBCAY/DQo="
    },
    {
      "t": 1674668017065,
      "m": "KgYBAgM2DQo="
    },
    {
      "t": 1674668017165,
      "m": "KgYBCQM9DQo="
    },
    {
      "t": 1674668017215,
      "m": "KgYBAwA0DQo="
    },
    {
      "t": 1674668017367,
      "m": "KgYBBwA4DQo="
    },
    {
      "t": 1674668017995,
      "m": "KgYBCQA6DQo="
    },
    {
      "t": 1674668018055,
      "m": "KgYBAgM2DQo="
    },
    {
      "t": 1674668018116,
      "m": "KgYBCQlDDQo="
    },
    {
      "t": 1674668018270,
      "m": "KgYBAwA0DQo="
    },
    {
      "t": 1674668018477,
      "m": "KgYBCQZADQo="
    },
    {
      "t": 1674668018538,
      "m": "KgYBAwk9DQo="
    },
    {
      "t": 1674668018759,
      "m": "KgYBCQM9DQo="
    },
    {
      "t": 1674668018838,
      "m": "KgYBAgAzDQo="
    },
    {
      "t": 1674668018973,
      "m": "KgYBCAY/DQo="
    },
    {
      "t": 1674668019015,
      "m": "KgYBAgM2DQo="
    },
    {
      "t": 1674668019135,
      "m": "KgYBCAlCDQo="
    },
    {
      "t": 1674668019231,
      "m": "KgYBCAA5DQo="
    }
  ]
}