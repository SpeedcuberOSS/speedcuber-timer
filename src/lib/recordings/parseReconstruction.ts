// Copyright (c) 2023 Joseph Hale <me@jhale.dev>
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import reconstructionFor, { SolutionMethod } from '../reconstructions';
import { STIF } from '../stif';
import {
  compressDoubleTurns,
  parseTimestampedMoves,
} from './parseTimestampedMoves';

export function parseReconstruction(
  recording: STIF.SolveRecording,
  scramble: STIF.Algorithm,
  startTime: number = 0,
  method: SolutionMethod = 'CFOP',
) {
  const moves = parseTimestampedMoves(recording, startTime);
  const rawReconstruction = reconstructionFor(scramble, moves).using(method);
  const reconstruction = rawReconstruction.map(phase => {
    return {
      label: phase.label,
      moves: compressDoubleTurns(phase.moves),
    };
  });
  return reconstruction;
}
