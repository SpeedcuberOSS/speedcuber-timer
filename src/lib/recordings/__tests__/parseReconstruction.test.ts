// Copyright (c) 2023 Joseph Hale <me@jhale.dev>
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import { STIF } from '../../stif';
import { parseReconstruction } from '../parseReconstruction';

const rcAttempt =
  require('../__fixtures__/rubiks_connected_attempt.json') as STIF.Attempt;
const rcRecording =
  require('../__fixtures__/rubiks_connected_recording.json') as STIF.SolveRecording;

describe('parseReconstruction', () => {
  it('returns no reconstruction if there is no message stream', () => {
    const recording = rcRecording;
    const actual = parseReconstruction(
      { smartPuzzle: recording.smartPuzzle, stream: [] },
      rcAttempt.solutions[0].scramble,
      rcAttempt.timerStart,
    );
    expect(actual).toEqual([]);
  });

  it('correctly parses a solve replay from a Rubiks Connected attempt', () => {
    const attempt = rcAttempt;
    const scramble = attempt.solutions[0].scramble;
    const actual = parseReconstruction(rcRecording, scramble, attempt.timerStart);
    expect(actual).toMatchSnapshot();
  });

  it('correctly parses a solve replay from a Particula 2x2x2', () => {
    const attempt =
      require('../__fixtures__/particula_2x2x2_attempt.json') as STIF.Attempt;
    const recording =
      require('../__fixtures__/particula_2x2x2_recording.json') as STIF.SolveRecording;
    const scramble = attempt.solutions[0].scramble;
    const actual = parseReconstruction(recording, scramble, attempt.timerStart);
    expect(actual).toMatchSnapshot();
  });

  it('correctly parses a solve replay from a Particula 3x3x3', () => {
    const attempt =
      require('../__fixtures__/particula_3x3x3_attempt.json') as STIF.Attempt;
    const recording =
      require('../__fixtures__/particula_3x3x3_recording.json') as STIF.SolveRecording;
    const scramble = attempt.solutions[0].scramble;
    const actual = parseReconstruction(recording, scramble, attempt.timerStart);
    expect(actual).toMatchSnapshot();
  });
});
