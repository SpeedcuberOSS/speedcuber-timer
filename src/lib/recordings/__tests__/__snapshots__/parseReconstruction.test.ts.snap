// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseReconstruction correctly parses a solve replay from a Particula 2x2x2 1`] = `
[
  {
    "label": "unknown",
    "moves": [
      {
        "m": "R",
        "t": 196,
      },
      {
        "m": "F",
        "t": 706,
      },
      {
        "m": "U",
        "t": 1247,
      },
      {
        "m": "F",
        "t": 1457,
      },
      {
        "m": "U'",
        "t": 1757,
      },
      {
        "m": "R",
        "t": 1856,
      },
      {
        "m": "U",
        "t": 2026,
      },
      {
        "m": "F'",
        "t": 2056,
      },
      {
        "m": "U'",
        "t": 2237,
      },
      {
        "m": "R",
        "t": 2567,
      },
      {
        "m": "U",
        "t": 2658,
      },
      {
        "m": "F2",
        "t": 2751,
      },
      {
        "m": "U'",
        "t": 2957,
      },
      {
        "m": "R2",
        "t": 4397,
      },
      {
        "m": "U2",
        "t": 5269,
      },
      {
        "m": "R'",
        "t": 5418,
      },
      {
        "m": "U2",
        "t": 5597,
      },
      {
        "m": "R2",
        "t": 5849,
      },
      {
        "m": "F2",
        "t": 6647,
      },
      {
        "m": "R'",
        "t": 6857,
      },
      {
        "m": "F2",
        "t": 6985,
      },
      {
        "m": "R",
        "t": 7217,
      },
    ],
  },
]
`;

exports[`parseReconstruction correctly parses a solve replay from a Particula 3x3x3 1`] = `
[
  {
    "label": "cross",
    "moves": [
      {
        "m": "B'",
        "t": 298,
      },
      {
        "m": "L",
        "t": 409,
      },
      {
        "m": "B",
        "t": 617,
      },
      {
        "m": "D'",
        "t": 779,
      },
      {
        "m": "U'",
        "t": 1050,
      },
    ],
  },
  {
    "label": "1st pair",
    "moves": [
      {
        "m": "B'",
        "t": 1651,
      },
      {
        "m": "D",
        "t": 1745,
      },
      {
        "m": "B",
        "t": 2015,
      },
    ],
  },
  {
    "label": "2nd pair",
    "moves": [
      {
        "m": "F'",
        "t": 2669,
      },
      {
        "m": "D2",
        "t": 2772,
      },
      {
        "m": "F",
        "t": 3000,
      },
      {
        "m": "D",
        "t": 3179,
      },
      {
        "m": "F'",
        "t": 3301,
      },
      {
        "m": "D'",
        "t": 3417,
      },
      {
        "m": "F",
        "t": 3576,
      },
    ],
  },
  {
    "label": "3rd pair",
    "moves": [
      {
        "m": "D'",
        "t": 3931,
      },
      {
        "m": "B",
        "t": 4027,
      },
      {
        "m": "D2",
        "t": 4292,
      },
      {
        "m": "B'",
        "t": 4509,
      },
      {
        "m": "D2",
        "t": 4587,
      },
      {
        "m": "B",
        "t": 4833,
      },
      {
        "m": "D'",
        "t": 4977,
      },
      {
        "m": "B",
        "t": 5235,
      },
      {
        "m": "B2",
        "t": 5309,
      },
    ],
  },
  {
    "label": "4th pair",
    "moves": [
      {
        "m": "F",
        "t": 5646,
      },
      {
        "m": "D'",
        "t": 5698,
      },
      {
        "m": "F'",
        "t": 5848,
      },
      {
        "m": "D",
        "t": 5910,
      },
      {
        "m": "R'",
        "t": 6719,
      },
      {
        "m": "D2",
        "t": 6864,
      },
      {
        "m": "R",
        "t": 7018,
      },
      {
        "m": "D",
        "t": 7137,
      },
      {
        "m": "R'",
        "t": 7172,
      },
      {
        "m": "D2",
        "t": 7259,
      },
      {
        "m": "R",
        "t": 7437,
      },
    ],
  },
  {
    "label": "OLLCP",
    "moves": [
      {
        "m": "D2",
        "t": 8010,
      },
      {
        "m": "B",
        "t": 8591,
      },
      {
        "m": "R",
        "t": 8703,
      },
      {
        "m": "D",
        "t": 8787,
      },
      {
        "m": "R'",
        "t": 8823,
      },
      {
        "m": "D'",
        "t": 8910,
      },
      {
        "m": "R",
        "t": 9027,
      },
      {
        "m": "D",
        "t": 9087,
      },
      {
        "m": "R'",
        "t": 9157,
      },
      {
        "m": "D'",
        "t": 9207,
      },
      {
        "m": "B'",
        "t": 9327,
      },
    ],
  },
  {
    "label": "EPLL",
    "moves": [
      {
        "m": "D2",
        "t": 9960,
      },
      {
        "m": "R'",
        "t": 10259,
      },
      {
        "m": "D",
        "t": 10319,
      },
      {
        "m": "R'",
        "t": 10365,
      },
      {
        "m": "D'",
        "t": 10482,
      },
      {
        "m": "R'",
        "t": 10648,
      },
      {
        "m": "D'",
        "t": 10678,
      },
      {
        "m": "R'",
        "t": 10767,
      },
      {
        "m": "D",
        "t": 10860,
      },
      {
        "m": "R",
        "t": 11038,
      },
      {
        "m": "D",
        "t": 11127,
      },
      {
        "m": "R2",
        "t": 11355,
      },
    ],
  },
  {
    "label": "AUF",
    "moves": [
      {
        "m": "D2",
        "t": 11496,
      },
    ],
  },
]
`;

exports[`parseReconstruction correctly parses a solve replay from a Rubiks Connected attempt 1`] = `
[
  {
    "label": "cross",
    "moves": [
      {
        "m": "L'",
        "t": 531,
      },
      {
        "m": "R'",
        "t": 616,
      },
      {
        "m": "B",
        "t": 1187,
      },
      {
        "m": "D'",
        "t": 1457,
      },
      {
        "m": "L",
        "t": 1684,
      },
      {
        "m": "U'",
        "t": 1947,
      },
      {
        "m": "B'",
        "t": 2626,
      },
    ],
  },
  {
    "label": "1st pair",
    "moves": [
      {
        "m": "F2",
        "t": 3288,
      },
      {
        "m": "D",
        "t": 3860,
      },
      {
        "m": "F'",
        "t": 4065,
      },
      {
        "m": "D'",
        "t": 4366,
      },
    ],
  },
  {
    "label": "2nd pair",
    "moves": [
      {
        "m": "D'",
        "t": 5176,
      },
      {
        "m": "F2",
        "t": 5298,
      },
      {
        "m": "D",
        "t": 5506,
      },
      {
        "m": "F",
        "t": 5598,
      },
      {
        "m": "U'",
        "t": 5817,
      },
      {
        "m": "F'",
        "t": 5986,
      },
      {
        "m": "U",
        "t": 6137,
      },
    ],
  },
  {
    "label": "3rd pair",
    "moves": [
      {
        "m": "D'",
        "t": 6412,
      },
      {
        "m": "F",
        "t": 6467,
      },
      {
        "m": "D",
        "t": 6676,
      },
      {
        "m": "R'",
        "t": 7187,
      },
      {
        "m": "F2",
        "t": 7312,
      },
      {
        "m": "R",
        "t": 7547,
      },
      {
        "m": "F'",
        "t": 7726,
      },
      {
        "m": "U",
        "t": 7880,
      },
      {
        "m": "F",
        "t": 8027,
      },
      {
        "m": "U'",
        "t": 8147,
      },
    ],
  },
  {
    "label": "4th pair",
    "moves": [
      {
        "m": "R",
        "t": 8536,
      },
      {
        "m": "F'",
        "t": 8631,
      },
      {
        "m": "R'",
        "t": 8716,
      },
      {
        "m": "D",
        "t": 8986,
      },
      {
        "m": "R'",
        "t": 9136,
      },
      {
        "m": "D'",
        "t": 9269,
      },
      {
        "m": "R",
        "t": 9316,
      },
    ],
  },
  {
    "label": "OLLCP",
    "moves": [
      {
        "m": "F2",
        "t": 9856,
      },
      {
        "m": "F",
        "t": 10465,
      },
      {
        "m": "U",
        "t": 10736,
      },
      {
        "m": "F",
        "t": 11031,
      },
      {
        "m": "L",
        "t": 11116,
      },
      {
        "m": "F'",
        "t": 11207,
      },
      {
        "m": "L'",
        "t": 11296,
      },
      {
        "m": "U'",
        "t": 11514,
      },
      {
        "m": "F'",
        "t": 11687,
      },
      {
        "m": "D",
        "t": 11972,
      },
      {
        "m": "R",
        "t": 12048,
      },
      {
        "m": "F",
        "t": 12196,
      },
      {
        "m": "R'",
        "t": 12296,
      },
      {
        "m": "F'",
        "t": 12346,
      },
      {
        "m": "D'",
        "t": 12498,
      },
    ],
  },
  {
    "label": "EPLL",
    "moves": [
      {
        "m": "R'",
        "t": 13126,
      },
      {
        "m": "F",
        "t": 13186,
      },
      {
        "m": "R'",
        "t": 13247,
      },
      {
        "m": "F'",
        "t": 13401,
      },
      {
        "m": "R'",
        "t": 13608,
      },
      {
        "m": "F'",
        "t": 13669,
      },
      {
        "m": "R'",
        "t": 13890,
      },
      {
        "m": "F",
        "t": 13969,
      },
      {
        "m": "R",
        "t": 14104,
      },
      {
        "m": "F",
        "t": 14146,
      },
      {
        "m": "R2",
        "t": 14266,
      },
    ],
  },
]
`;
